### PR TITLE
IO-649 - Improve the performance of the contentEquals() methods.

### PR DESCRIPTION
**Pull request by brettlounsbury**
This change modifies the contentEquals() methods to internally buffer content into a byte/char array and to then do batch comparisons of those arrays using Arrays.equals instead of using a BufferedInputStream or BufferedReader and making use of the single byte/char read() methods.

This reduces the number of method invocations by a factor equal to the buffer size and avoids casting every byte read to an int and improves performance significantly.

The following table shows the performance increase over 1000 iterations of comparing 2 1GB InputStream of binary data (stored in memory to avoid I/O).  This test was performed on an EC2 M4.4XL host using Java 1.8.0.232 and there was a forced System.gc() between each iteration to avoid GC as a source of latency:

Average: 7236 to 858ms (8.43x speedup)
P50: 7224 to 856ms (8.44x speedup)
P90: 7249 to 860ms (8.43x speedup)
P99: 7410 to 913ms (8.12x speedup)
P100: 8330 to 1278ms (6.52x speedup)

The following table shows the performance increase over 1000 iterations of comparing 2 1GB Reader of character data (stored in memory to avoid I/O).  This test was performed on an EC2 M4.4XL host using Java 1.8.0.232 and there was a forced System.gc() between each iteration to avoid GC as a source of latency:

Average: 11281 to 1737ms (6.50x speedup)
P50: 11262 to 1735ms (6.49x speedup)
P90: 11292 to 1741ms (6.49x speedup)
P99: 11707 to 1774ms (6.60x speedup)
P100: 12176 to 1884ms (6.46x speedup)

### CODE REVIEW

**Comment by brettlounsbury** [comment]
https://issues.apache.org/jira/browse/IO-649

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 1]
```diff
@@ -708,31 +706,79 @@ public static void closeQuietly(final Writer output) {
...
+     * @return true if the content of the streams are equal or they both don't
+     * exist, false otherwise
+     * @throws NullPointerException if either input is null
+     * @throws IOException          if an I/O error occurs
+     */
+    @SuppressWarnings("resource")
+    public static boolean contentEquals(final InputStream input1, final InputStream input2, final int bufferSize)
+            throws IOException {
+        if (bufferSize <= 0) {
+            throw new IllegalArgumentException("Buffer size must be positive: " + bufferSize);
```
Comment (michael-o) [inline src/main/java/org/apache/commons/io/IOUtils.java]: This one is not documented.

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 2]
```diff
@@ -708,31 +706,79 @@ public static void closeQuietly(final Writer output) {
...
+     * Compares the contents of two Streams to determine if they are equal or not.
+     * <p>
+     * This method buffers the input internally.
+     *
+     * @param input1 the first stream
+     * @param input2 the second stream
+     * @param bufferSize the size of the internal buffer to use.
+     * @return true if the content of the streams are equal or they both don't
+     * exist, false otherwise
+     * @throws NullPointerException if either input is null
```
Comment (michael-o) [inline src/main/java/org/apache/commons/io/IOUtils.java]: I fail to see where this one is thrown.

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 3]
```diff
@@ -746,25 +792,74 @@ public static boolean contentEquals(final InputStream input1, final InputStream
...
+     * <p>
+     * This method buffers the input internally.
+     * </p>
+     *
+     * @param input1 the first reader
+     * @param input2 the second reader
+     * @param bufferSize the size of the internal buffer to use.
+     * @return true if the content of the readers are equal or they both don't
+     * exist, false otherwise
+     * @throws NullPointerException if either input is null
```
Comment (michael-o) [inline src/main/java/org/apache/commons/io/IOUtils.java]: Same here.

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 2 above)
Comment (brettlounsbury) [reply src/main/java/org/apache/commons/io/IOUtils.java]: This documentation is a clone/mutate from the original method's javadoc (the one without the bufferSize).  This doesn't look to be explicitly thrown, but if either InputStream is null it throws due to the stream being dereferenced.

Check the javadoc for the `public static boolean contentEquals(final InputStream input1, final InputStream input2)` method.

Happy to either (a) explicitly throw a NullPointerException up front, (b) remove the comment from both places, or (c) do nothing.

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 3 above)
Comment (brettlounsbury) [reply src/main/java/org/apache/commons/io/IOUtils.java]: Same answer as above :-)

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 1 above)
Comment (brettlounsbury) [reply src/main/java/org/apache/commons/io/IOUtils.java]: Fixed and updated.  Done for both methods that throw an IllegalArgumentException.

File: src/main/java/org/apache/commons/io/IOUtils.java (hunk 2 above)
Comment (michael-o) [reply src/main/java/org/apache/commons/io/IOUtils.java]: I would rather see an explicit NPE up front.

Comment (brettlounsbury) [reply src/main/java/org/apache/commons/io/IOUtils.java]: I looked at the beginning of each method again.  Both contentEquals methods start with the following statements.  This can never throw a NullPointerException.  If both streams are null they will be equal and therefore always return true.  If one stream is null and the other is not it will always return false based on the XOR logic.  Only if both streams are non-null and not the same object will the actual logic of the method execute.

```
if (input1 == input2) {
            return true;
}
if (input1 == null ^ input2 == null) {
            return false;
}
```

**Comment by brettlounsbury** [comment]
I took a look at the coverage drop.  I forgot to add tests that exercise the bounds checking of the bufferSize inputs to both contentEquals method.  I've pushed a new version with that.

**Comment by brettlounsbury** [comment]
@garydgregory - can you take a look or propose another reviewer?

**Comment by garydgregory** [comment]
Hi @brettlounsbury.

Thank you for your PR and updates. I've sprinkled comments throughout.

Performance claims must be backed up somehow, please add a JMH benchmark to the test so myself and others can compare mileage. 

For examples of JMH usage, please see other Apache Commons components like Commons BCEL, Commons Crypto, Commons CSV, Commons Lang, Commons Math, and Commons RNG.

Thank you!
Gary

**Comment by brettlounsbury** [comment]
Hi @garydgregory,

I don't see any comments from you other than the request for a benchmark. 

I have updated IOUtils to include a test dependency on JMH and and to include a benchmark that shows BufferedInputStream.read() performance compared to InputStream.read(byte[]) performance.  This test creates an 8MB file on disk once before the benchmark is ran and reads it back in.  It then just keeps a running sum of the bytes in the file (and ignores any overflow/underflow) and blackhole's the sum at the end to avoid any compiler optimization.

Results look something like this:
Benchmark                                                        Mode  Cnt         Score        Error  Units
SingleByteReadVsMultiByteReadPerformanceTest.testReadMultiBytes  avgt   25   6220618.290 ± 204562.812  ns/op
SingleByteReadVsMultiByteReadPerformanceTest.testReadSingleByte  avgt   25  21740409.146 ± 711294.924  ns/op

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 4]
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
...
+     * returns the buffer will either be full or the InputStream will be fully consumed.
+     *
+     * @param inputStream the InputStream from which to read
+     * @param buffer the buffer to fill
+     * @return the number of bytes placed in the buffer.  If this is equal to <code>buffer.length</code> then the
+     * buffer was completely filled.  If it is less than <code>buffer.length</code>, then the InputStream has been
+     * fully consumed.
+     * @throws IOException if an I/O error occurs.
+     */
+    private static int fillBuffer(final InputStream inputStream, final byte[] buffer) throws IOException {
```
[the remark this answers is not in the document]
Comment (brettlounsbury) [reply src/main/java/org/apache/commons/io/IOUtils.java]: Yes, it looks like IOUtils.read(InputStream, byte[]) and IOUtils.read(Reader, char[]) (without the optional arguments) should do exactly what I want.

Thanks for the call out, I didn't realize those methods existed.

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 5]
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
...
      * @throws IOException          if an I/O error occurs
      */
     @SuppressWarnings("resource")
     public static boolean contentEquals(final InputStream input1, final InputStream input2)
             throws IOException {
+        return contentEquals(input1, input2, DEFAULT_BUFFER_SIZE);
+    }
+
+    /**
+     * Compares the contents of two Streams to determine if they are equal or not.
```
[the remark this answers is not in the document]
Comment (brettlounsbury) [reply src/main/java/org/apache/commons/io/IOUtils.java]: Done.

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 6]
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
...
+     * This method buffers the input internally.
+     *
+     * @param input1 the first stream
+     * @param input2 the second stream
+     * @param bufferSize the size of the internal buffer to use.
+     * @return true if the content of the streams are equal or they both don't
+     * exist, false otherwise
+     * @throws IllegalArgumentException if bufferSize is less than or equal to zero.
+     * @throws IOException              if an I/O error occurs
+     * @since 2.6
```
[the remark this answers is not in the document]
Comment (brettlounsbury) [reply src/main/java/org/apache/commons/io/IOUtils.java]: Done

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 7]
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
...
+     * fully consumed.
+     * @throws IOException if an I/O error occurs.
+     */
+    private static int fillBuffer(final InputStream inputStream, final byte[] buffer) throws IOException {
+        int totalBytesRead = 0;
+        int bytesRead = 0;
+        do {
+            totalBytesRead += bytesRead;
+            bytesRead = inputStream.read(buffer, totalBytesRead, buffer.length - totalBytesRead);
+        } while (bytesRead != -1 && totalBytesRead < buffer.length);
```
[the remark this answers is not in the document]
Comment (brettlounsbury) [reply src/main/java/org/apache/commons/io/IOUtils.java]: Good catch.  I've used it elsewhere but somehow missed it in the loops.  Its somewhat moot, however, since the two fillBuffer methods have been deleted.

File: src/main/java/org/apache/commons/io/IOUtils.java [hunk 8]
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
...
+     * fully consumed.
+     * @throws IOException if an I/O error occurs.
+     */
+    private static int fillBuffer(final Reader reader, final char[] buffer) throws IOException {
+        int totalBytesRead = 0;
+        int bytesRead = 0;
+        do {
+            totalBytesRead += bytesRead;
+            bytesRead = reader.read(buffer, totalBytesRead, buffer.length - totalBytesRead);
+        } while (bytesRead != -1 && totalBytesRead < buffer.length);
```
[the remark this answers is not in the document]
Comment (brettlounsbury) [reply src/main/java/org/apache/commons/io/IOUtils.java]: Same as above :-)

**Comment by brettlounsbury** [comment]
@garydgregory,

I really appreciate your thorough review of my code.  I think at this point I have addressed your comments, but if you have any others please let me know.

I will add that I made the IOUtils.DEFAULT_BUFFER_SIZE constant package private (default visibility) so it could be referenced from IOUtils tests so that I can ensure the logic makes sense around buffer boundaries and I didn't want to clone that property since it could later be mutated and invalidate some of the edge condition testing being done.

**Comment by brettlounsbury** [comment]
@garydgregory - Any other feedback here or anything else I can do to improve this?

**Comment by brettlounsbury** [comment]
@garydgregory - Can you take another look or suggest another reviewer?

**Comment by garydgregory** [comment]
Sorry about the delay, last day of vacation here ;-)

**Comment by brettlounsbury** [comment]
> Sorry about the delay, last day of vacation here ;-)

No worries - hope you had a great vacation!

**Comment by brettlounsbury** [comment]
@garydgregory - can you take another look when you have a chance.

**Comment by brettlounsbury** [comment]
@garydgregory - I think all necessary changes have been made.  Are you able to take a look?

**Comment by garydgregory** [comment]
@brettlounsbury May you please rebase on master?

**Comment by brettlounsbury** [comment]
@garydgregory - resolved the merge conflict, should be good now.

**Comment by garydgregory** [comment]
Closing, no feedback.
